File: frontend/src/types/forecast.ts
export interface ForecastRequest {
  address: string
}

export interface LocationInfo {
  label: string
  latitude: number
  longitude: number
  timezone: string
}

export interface TimeWindow {
  start: string
  end: string
}

export interface BestHourWindow {
  start: string
  end: string
  score: number
}

export interface CloudCoverBreakdown {
  total: number | null
  low: number | null
  mid: number | null
  high: number | null
}

export interface PrecipitationBreakdown {
  total_mm: number | null
  max_hourly_mm: number | null
  max_probability: number | null
}

export interface HourlyScore {
  time: string
  at: string
  score: number
  moon_illumination_effective?: number | null
  moon_up?: boolean | null
  moon_altitude?: number | null
  cloud_cover?: number | null
  cloud_cover_low?: number | null
  cloud_cover_mid?: number | null
  cloud_cover_high?: number | null
  visibility?: number | null
  seeing?: number | null
  transparency?: number | null
  precipitation?: number | null
  precipitation_probability?: number | null
  dew_point?: number | null
  temperature?: number | null
}

export interface MeteorShowerHighlight {
  id: string
  name: string
  zhr_nominal: number | null
}

export interface NightForecast {
  date: string
  rating: string
  score: number | null
  moon_phase: string
  moon_illumination: number
  moonrise?: string | null
  moonset?: string | null
  moon_sky_glow_avg?: number | null
  temperature_high?: number | null
  temperature_low?: number | null
  cloud_cover: CloudCoverBreakdown
  precipitation: PrecipitationBreakdown
  dark_window?: TimeWindow | null
  best_hours: BestHourWindow[]
  hourly: HourlyScore[]
  no_darkness: boolean
  meteor_showers: MeteorShowerHighlight[]
  astro_forecast_limited?: boolean
}

export interface ForecastResponse {
  location: LocationInfo
  nights: NightForecast[]
  score_step_minutes?: number
  prior_day_dark_window?: TimeWindow | null
}

File: backend/app/models/forecast.py
"""Pydantic schemas for the forecast API."""

from pydantic import BaseModel, Field


class ForecastRequest(BaseModel):
    address: str = Field(..., min_length=1, max_length=500)


class LocationInfo(BaseModel):
    label: str
    latitude: float
    longitude: float
    timezone: str


class TimeWindow(BaseModel):
    start: str
    end: str


class BestHourWindow(BaseModel):
    start: str
    end: str
    score: int


class CloudCoverBreakdown(BaseModel):
    total: float | None = None
    low: float | None = None
    mid: float | None = None
    high: float | None = None


class PrecipitationBreakdown(BaseModel):
    total_mm: float | None = None
    max_hourly_mm: float | None = None
    max_probability: float | None = None


class HourlyScore(BaseModel):
    time: str
    at: str
    score: int
    moon_illumination_effective: float | None = None
    moon_up: bool | None = None
    moon_altitude: float | None = None
    cloud_cover: float | None = None
    cloud_cover_low: float | None = None
    cloud_cover_mid: float | None = None
    cloud_cover_high: float | None = None
    visibility: float | None = None
    seeing: int | None = None
    transparency: int | None = None
    precipitation: float | None = None
    precipitation_probability: float | None = None
    dew_point: float | None = None
    temperature: float | None = None


class MeteorShowerHighlight(BaseModel):
    id: str
    name: str
    zhr_nominal: int | None = None


class NightForecast(BaseModel):
    date: str
    rating: str
    score: int | None
    moon_phase: str
    moon_illumination: float
    moonrise: str | None = None
    moonset: str | None = None
    moon_sky_glow_avg: float | None = None
    temperature_high: float | None = None
    temperature_low: float | None = None
    cloud_cover: CloudCoverBreakdown = Field(default_factory=CloudCoverBreakdown)
    precipitation: PrecipitationBreakdown = Field(default_factory=PrecipitationBreakdown)
    dark_window: TimeWindow | None = None
    best_hours: list[BestHourWindow] = Field(default_factory=list)
    hourly: list[HourlyScore] = Field(default_factory=list)
    no_darkness: bool = False
    meteor_showers: list[MeteorShowerHighlight] = Field(default_factory=list)
    astro_forecast_limited: bool = True


class ForecastResponse(BaseModel):
    location: LocationInfo
    nights: list[NightForecast]
    score_step_minutes: int = 60
    prior_day_dark_window: TimeWindow | None = None
    astro_data_unavailable: bool = False
